### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by timcdlucas**
Hello,

Firstly, thank you for making this code so open, well organised and well documented. It's really appreciated.

I've just started working with Petra via the RAMP programme. We're talking about a project on assessing the effectiveness of contract tracing as we release lockdown and other measures. So we're looking at using this code as a starting point. As the project progresses I imagine I may well talk to you further either about the model or about shared authorship if the analysis ends up relying strongly on your code.

As a way to get to know the code better I started writing unit tests and making some other edits. I can imagine you might *not* want to merge them into this repo as the repo is so closely linked to the lancet paper. But a PR seemed an appropriate way to say hello and let you see the tests that I have written. If you're still working on similar models perhaps these unit tests would be useful in a different repo for example. I don't know.

I intend to continue working with this code. I may refactor the code and then write more tests for example. Happy to PR any further work to this repo or a similar repo if that's useful.

Anyway, thanks again,
Tim

### CODE REVIEW

File: .gitignore
Comment (seabbs) [inline .gitignore:1]: Not clear why the gitignore has been reset?

File: R/figures2to4.R
Comment (seabbs) [inline R/figures2to4.R:244]: typo

File: R/figures2to4.R
Comment (seabbs) [inline R/figures2to4.R:129]: More clarity here.

File: R/figures2to4.R
Comment (seabbs) [reply R/figures2to4.R:244]: More clarity on the expected input

File: R/outbreak_model.R
Comment (seabbs) [inline R/outbreak_model.R:38]: We use the NULLs to stop functions calling things from the global environment. This helps give clarity over if the function is working. Please put back. 

Also prop.ascertainL -> prop.ascertain

File: R/outbreak_step.R
Comment (seabbs) [inline R/outbreak_step.R:34]: As above

File: R/outbreak_step.R
Comment (seabbs) [inline R/outbreak_step.R:152]: suggest replace with pmax, pmin

**Comment by jhellewell14** [comment]
Hello Tim, nice to meet you!

We (me and @seabbs) have already been discussing the potential of using this code to answer the after-lockdown contact tracing question but decided that we have too much work to do. So it's nice to see someone else picking this code up. 

I'm very happy for you to build on the package and would be keen to support the work going forward. 

We have a "paper" version of the repository hosted somewhere else so I'm happy to merge these changes (pending a review by @seabbs) here and we will put the previous code as a github release and you can build on top of it.

**Comment by timcdlucas** [comment]
Hi both,

"It would be good to see more extensive and unit based tests."
Totally agree. But I think I'll have to refactor the code to do that. I think I will do that, but it will take a while.

"decided that we have too much work to do."
I guess this is exactly what RAMP is for! If you were already thinking about it it makes sense for you to be as involved as you wish with me/others in my lab doing the leg work.

I'll ask some further questions in the specific commit comments. But I know you're busy so please feel no rush to deal with any of this. I'll happily carry on working on my fork anyway.

File: R/outbreak_model.R
Comment (timcdlucas) [reply R/outbreak_model.R:38]: I don't want to take up your time so I will put this back as requested.

However, I don't see the logic here. If there is no default, the function won't call anything from the global environment, it will just give an informative error.

```
f <- function(x) x ^ 2
x <- 4
f()
```

I guess if you mean when you're running the code line by line to debug and you always start by setting the parameters to default?

Comment (seabbs) [reply R/outbreak_model.R:38]: Yes that is it :smile: I am not clear what the upside is from not setting it?

**Comment by seabbs** [comment]
Hi Tim, 

Great to have you reviewing/improving the code/expanding the research. As Joels says this is very much in the direction we were thinking of looking (if I understand what you are looking at) so sounds like a good idea to join forces! 

In terms of unit tests, I agree more modularity in the actual functions would help for testing but think even with the current set up it is possible to test that certain behaviour does what it says on the tin (not that you need to write these tests). 

Sam

File: R/outbreak_model.R
Comment (timcdlucas) [reply R/outbreak_model.R:38]: The upside is that if you run a function without a needed argument, it immediately errors and tells you which argument you missed. Rather than trying to run the code with the arg set as NULL and seeing what happens.

Here's a contrived example just to be clear. The second function is doesn't error and probably isn't doing what you want it to do.

**Comment by sbfnk** [comment]
@timcdlucas it may have taken a few years but I we're currently looking into splitting up the repo into a paper repo and an eventually more general R package - for which these tests are very useful! I've polished things a bit and added you as a package co-author. I hope you're ok with this, please do shout if not.

**Review by sbfnk — APPROVED** [review]

**Comment by timcdlucas** [comment]
Totally fine by me. It was an interesting flash back to see these pop up in my email!
Tim

**Comment by seabbs** [comment]
Woo yes a blast from the past. Thanks again for writing these @timcdlucas and sorry they didn't get merged in sooner!

**Pull request merged**
